server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAutoShardingEvent.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.SimpleDiffable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.function.LongSupplier;
+
+/**
+ * Represents the last auto sharding event that occured for a data stream.
+ */
+public record DataStreamAutoShardingEvent(String triggerIndexName, long generation, int targetNumberOfShards, long timestamp)
```

Comment:
We probably need a `DataStreamAutoShardingEventTests` class that extends `SimpleDiffableSerializationTestCase` to check that everything is good with serialization of this class